File: Books/Array/JS/Sorts/selectSort.js
class SelectSortArray {
    a;
    nElems;

    constructor(size) {
        this.a = new Array(size);
        this.nElems = 0;
    }

    insert(value) {
        this.a[this.nElems] = value;
        this.nElems++;
    }

    find(searchKey) {
        let lowerBound = 0;
        let upperBound = this.nElems - 1;
        let curInd;

        while (true) {
            curInd = Math.floor((lowerBound + upperBound) / 2);

            if (lowerBound > upperBound) {
                console.log("Can't find " + searchKey);
                return this.nElems;
            }

            if (this.a[curInd] === searchKey) {
                console.log("Found " + searchKey);
                return curInd;
            } else {
                if (this.a[curInd] < searchKey) {
                    lowerBound = curInd + 1;
                } else {
                    upperBound = curInd - 1;
                }
            }
        }
    }

    delete(value) {
        let i = this.find(value);

        if (i === this.nElems) {
            return false
        } else {
            for (let j = i; j < this.nElems; j++) {
                this.a[j] = this.a[j + 1];
            }
            this.nElems--;
        }
    }

    selectSort() {
        let out;
        let inner;
        let min;

        for (out = 0; out < this.nElems - 1; out++) {
            min = out;

            for(inner = out + 1; inner < this.nElems; inner++) {
                if(this.a[inner] < this.a[min]) {
                    min = inner;
                    let temp = this.a[out];
                    this.a[out] = this.a[min];
                    this.a[min] = temp;
                }
            }
        }

    }

    display() {
        const list = [];
        for (let i = 0; i < this.nElems; i++) {
            list.push(this.a[i] + "")
        }
        console.log(list.join(' '));
    }

}

function SelectSortApp() {
    const maxSize = 100;
    const arr = new SelectSortArray(maxSize)

    arr.insert(44);
    arr.insert(88);
    arr.insert(33);
    arr.insert(66);
    arr.insert(99);
    arr.insert(55);
    arr.insert(22);
    arr.insert(77);
    arr.insert(11);

    arr.display();

    arr.selectSort();

    arr.display();

}

SelectSortApp();

File: Books/Array/Java/Sorts/selectSort.java
class SelectSortApp {
    public static void main(String[] args) {

        int maxSize = 100;
        SelectSortArray arr;

        arr = new SelectSortArray(maxSize);

        arr.insert(44);
        arr.insert(88);
        arr.insert(33);
        arr.insert(66);
        arr.insert(99);
        arr.insert(55);
        arr.insert(22);
        arr.insert(77);
        arr.insert(11);

        arr.display();

        arr.selectSort();

        arr.display();

    }
}

class SelectSortArray {
    private long[] a;
    private int nElems;

    public SelectSortArray(int max) {
        a = new long[max];
        nElems = 0;
    }

    public void insert(long value) {
        a[nElems] = value;
        nElems++;
    }

    public int find(long searchKey) {
        int lowerBound = 0;
        int upperBound = nElems - 1;
        int curInd;

        while (true) {
            curInd = (lowerBound + upperBound) / 2;

            if (a[curInd] == searchKey) {
                System.out.println("Found " + searchKey);
                return curInd;
            } else if (lowerBound > upperBound) {
                System.out.println("Can't find " + searchKey);
                return nElems;
            } else {
                if (a[curInd] < searchKey) {
                    lowerBound = curInd + 1;
                } else {
                    upperBound = curInd - 1;
                }
            }
        }
    }

    public boolean delete(long value) {
        int j = find(value);
        if (j == nElems) {
            return false;
        } else {
            for (int k = j; k < nElems; k++) {
                a[k] = a[k + 1];
            }
            nElems--;
            return true;
        }
    }

    public void selectSort() {
        int out, in, min;
        for (out = 0; out < nElems - 1; out++) {
            min = out;
            for (in = out + 1; in < nElems; in++) {
                if(a[in] < a[min]) {
                    min = in;
                    long temp = a[out];
                    a[out] = a[min];
                    a[min] = temp;
                }
            }
        }
    }

    public void display() {
        for (int j = 0; j < nElems; j++)
            System.out.print(a[j] + " ");
        System.out.println("");
    }
};
